File: tests/test_caster_scope_1.cpp
#include "pybind11_tests.h"
#include "test_caster_scope.h"

namespace pybind11 {
namespace detail {
template <>
struct type_caster<test_caster_scope::atyp> : test_caster_scope::type_caster_atyp<1> {};
} // namespace detail
} // namespace pybind11

TEST_SUBMODULE(caster_scope_1, m) {
    using namespace test_caster_scope;

    m.def("rtrn_bval", rtrn_bval);
    m.def("rtrn_rref", rtrn_rref);
    m.def("rtrn_cref", rtrn_cref);
    m.def("rtrn_mref", rtrn_mref);
    m.def("rtrn_cptr", rtrn_cptr);
    m.def("rtrn_mptr", rtrn_mptr);

    m.def("pass_bval", pass_bval);
    m.def("pass_rref", pass_rref);
    m.def("pass_cref", pass_cref);
    m.def("pass_mref", pass_mref);
    m.def("pass_cptr", pass_cptr);
    m.def("pass_mptr", pass_mptr);
}

File: tests/test_caster_scope_1.py
import re

import pytest

from pybind11_tests import caster_scope_1 as m


@pytest.mark.parametrize(
    "rtrn_f, expected",
    [
        (m.rtrn_bval, "1cast_rref:bval_CpCtor"),
        (m.rtrn_rref, "1cast_rref:rref"),
        (m.rtrn_cref, "1cast_cref:cref"),
        (m.rtrn_mref, "1cast_mref:mref"),
        (m.rtrn_cptr, "1cast_cptr:cptr"),
        (m.rtrn_mptr, "1cast_mptr:mptr"),
    ],
)
def test_cast(rtrn_f, expected):
    assert re.match(expected, rtrn_f())


@pytest.mark.parametrize(
    "pass_f, expected",
    [
        (m.pass_bval, "pass_bval:1rref_MvCtor"),
        (m.pass_rref, "pass_rref:1rref"),
        (m.pass_cref, "pass_cref:1cref"),
        (m.pass_mref, "pass_mref:1mref"),
        (m.pass_cptr, "pass_cptr:1cptr"),
        (m.pass_mptr, "pass_mptr:1mptr"),
    ],
)
def test_operator(pass_f, expected):
    assert re.match(expected, pass_f(None))
